### PR TITLE
change envoy docker image to "distroless" base image.

### PR DESCRIPTION
**Pull request by phil9909**
Relates #12122 and #12694.
As requested in https://github.com/istio/istio/issues/12122#issuecomment-476109598

Check how integration tests and the distroless base image work together.

### CODE REVIEW

File: pilot/docker/Dockerfile.proxyv2 [hunk 1]
```diff
@@ -24,21 +19,7 @@ ADD envoy_policy.yaml.tmpl /etc/istio/proxy/envoy_policy.yaml.tmpl
 ADD envoy_telemetry.yaml.tmpl /etc/istio/proxy/envoy_telemetry.yaml.tmpl
 ADD istio-iptables.sh /usr/local/bin/istio-iptables.sh
 
-# Copy Envoy bootstrap templates used by pilot-agent
 COPY envoy_bootstrap_v2.json /var/lib/istio/envoy/envoy_bootstrap_tmpl.json
-COPY envoy_bootstrap_drain.json /var/lib/istio/envoy/envoy_bootstrap_drain.json
```
Comment (sdake) [inline pilot/docker/Dockerfile.proxyv2]: not sure why this line is deleted - could you expand?

File: pilot/docker/Dockerfile.proxyv2 [hunk 2]
```diff
@@ -24,21 +19,7 @@ ADD envoy_policy.yaml.tmpl /etc/istio/proxy/envoy_policy.yaml.tmpl
...
-COPY gcp_envoy_bootstrap.json /var/lib/istio/envoy/gcp_envoy_bootstrap_tmpl.json
```
Comment (sdake) [inline pilot/docker/Dockerfile.proxyv2]: ditto.

File: pilot/docker/Dockerfile.proxyv2 [hunk 3]
```diff
@@ -24,21 +19,7 @@ ADD envoy_policy.yaml.tmpl /etc/istio/proxy/envoy_policy.yaml.tmpl
...
-
-# Add CA certificates for SSL connections.
-# obtained from debian ca-certs deb using fetch_cacerts.sh
-ADD ca-certificates.tgz /
```
Comment (sdake) [inline pilot/docker/Dockerfile.proxyv2]: this makes sense as distroless containers include ca certificates (this is one thing that from scratch fails to produce).

File: pilot/docker/Dockerfile.proxyv2 [hunk 4]
```diff
@@ -24,21 +19,7 @@ ADD envoy_policy.yaml.tmpl /etc/istio/proxy/envoy_policy.yaml.tmpl
...
-
-RUN chmod 755 /usr/local/bin/envoy /usr/local/bin/pilot-agent
```
Comment (sdake) [inline pilot/docker/Dockerfile.proxyv2]: this line is not needed in the current dockerfile - could be removed in a separate PR while this pr goes into debate mode :)

File: pilot/docker/Dockerfile.proxyv2 [hunk 5]
```diff
@@ -24,21 +19,7 @@ ADD envoy_policy.yaml.tmpl /etc/istio/proxy/envoy_policy.yaml.tmpl
 ADD envoy_telemetry.yaml.tmpl /etc/istio/proxy/envoy_telemetry.yaml.tmpl
 ADD istio-iptables.sh /usr/local/bin/istio-iptables.sh
```
Comment (sdake) [inline pilot/docker/Dockerfile.proxyv2]: what precisely does istio-iptables.sh do?  If its a bash script, does it not need bash to operate?  Is it used in the implementation?

File: pilot/docker/Dockerfile.proxyv2 (hunk 1 above)
Comment (phil9909) [reply pilot/docker/Dockerfile.proxyv2]: Sorry, this was due to a outdated master.
We fixed this.

File: pilot/docker/Dockerfile.proxyv2 (hunk 2 above)
Comment (phil9909) [reply pilot/docker/Dockerfile.proxyv2]: Sorry, this was due to a outdated master.
We fixed this.

File: pilot/docker/Dockerfile.proxyv2 (hunk 4 above)
Comment (phil9909) [reply pilot/docker/Dockerfile.proxyv2]: done: https://github.com/istio/istio/pull/12745

**Comment by phil9909** [comment]
The test `TestCoreDumpGenerated` of the suite `prow/istio-pilot-e2e-envoyv2-v1alpha3.sh` is incompatible with the distroless base image.
It does the following:
1. copy a "crashing program" to the container (`kill -SIGSEGV $$` in a bash)
1. execute the crashing program
1. uses `find` to check if a coredump was generated
1. deletes the coredump

This does not work because the distroless image misses the following binaries:
* bash
* find
* kill
* rm
* tar (this is internally used by `kubectl cp`)

Adding all those binaries to the image would miss the point of building a distroless image.
A compromise would be adding a tar binary to restore `kubectl cp` functionality and rewriting the test to remove the dependencies on those other binaries.

**Comment by c0d1ngm0nk3y** [comment]
All other tests were at least once ok. We focused on the `required`, but we haven't seen a obvious problem in the others as well. But since the test stability is *FAR* from good, it is easy to miss something.

But other than the issue mentioned earlier, it seems that there is no test issue concerning a distroless envoy.

**Comment by c0d1ngm0nk3y** [comment]
> The test `TestCoreDumpGenerated` of the suite `prow/istio-pilot-e2e-envoyv2-v1alpha3.sh` is incompatible with the distroless base image....

The test aside, the fact that `kubectl cp` does not work anymore since it needs `tar` on the container, might be worth 'fixing', right?

**Comment by c0d1ngm0nk3y** [comment]
> The test `TestCoreDumpGenerated` of the suite `prow/istio-pilot-e2e-envoyv2-v1alpha3.sh` ...

The suite `prow/istio_auth_sds_e2e` fails with the same test `TestCoreDumpGenerated` . So this test is executed twice? But it is the same problem.

**Comment by robertpanzer** [comment]
Would these images still contain libraries like `libnettle`, `libhogweed` etc?
We are currently calling these from Lua filters in Istio because istio/proxy doesn't expose the embedded BoringSSL library.

If neither one nor the other is available, that would be a serious problem for us.

**Comment by sdake** [comment]
@robertpanzer - my take on this work is that we would end with two artifacts - the distro based images, and the distroless based images.  As to your question, I am not certain - that would require more testing.

Cheers
-steve

**Comment by sdake** [comment]
for future reference nettle: https://ftp.gnu.org/gnu/nettle/ contains libhogweed and libnettle

**Comment by sdake** [comment]
> The test `TestCoreDumpGenerated` of the suite `prow/istio-pilot-e2e-envoyv2-v1alpha3.sh` is incompatible with the distroless base image.
> It does the following:
> 
> 1. copy a "crashing program" to the container (`kill -SIGSEGV $$` in a bash)
> 2. execute the crashing program
> 3. uses `find` to check if a coredump was generated
> 4. deletes the coredump
> 
> This does not work because the distroless image misses the following binaries:
> 
> * bash
> * find
> * kill
> * rm
> * tar (this is internally used by `kubectl cp`)
> 
> Adding all those binaries to the image would miss the point of building a distroless image.
> A compromise would be adding a tar binary to restore `kubectl cp` functionality and rewriting the test to remove the dependencies on those other binaries.

Interesting that kubectl has a third party dependency on tar.  Wasn't aware of that.  It seems as though we may need some extra binary packages to meet the requirements of our community.  If we can get more exposure as to what these requirements are (beyond a full blown distro) it would help understand if this work is viable or would come with some limitations (e.g. distroless containers would lose some lua integration features.

Cheers
-steve

**Comment by c0d1ngm0nk3y** [comment]
This is also related: #12746 - The goal would be to offer a distroless variant of every image (at least as alternative).

**Comment by c0d1ngm0nk3y** [comment]
The ubuntu based images: #12788 - so we have a running version of all images and can discuss how to go on with that.

**Comment by c0d1ngm0nk3y** [comment]
With #12869 also the proxy_init works with distroless. But that one needs a bit extra work.

File: pilot/docker/Dockerfile.proxyv2 [hunk 6]
```diff
@@ -1,4 +1,4 @@
-FROM istionightly/base_debug
+FROM gcr.io/distroless/cc
```
Comment (costinm) [inline pilot/docker/Dockerfile.proxyv2]: /hold - please instead change base_debug to be build from distroless, and have all istio images be based on the base.

File: install/kubernetes/helm/istio/charts/security/templates/tests/test-citadel-connection.yaml [hunk 7]
```diff
@@ -20,7 +20,7 @@ spec:
 {{- end }}
   containers:
     - name: "{{ template "security.fullname" . }}-test"
-      image: {{ .Values.global.hub }}/{{ .Values.global.proxy.image }}:{{ .Values.global.tag }}
+      image: istionightly/base_debug
```
Comment (howardjohn) [inline install/kubernetes/helm/istio/charts/security/templates/tests/test-citadel-connection.yaml]: This doesn't look right? Is this supposed to be changed?

File: pilot/docker/Dockerfile.proxyv2 (hunk 6 above)
Comment (sdake) [reply pilot/docker/Dockerfile.proxyv2]: @costinm that won't work.  If you look deeper in the PRs, you will see that several files are copied in (such as the ca certs) to our existing images during the build process and untarred.

I also think there have been some communication problems here.  This was specifically an experiment I asked to be done to test - across Istio - if the gates could be made to work with distroless.

My thinking was not to replace base as you propose, but instead, have a second set of docker files specifically tuned for distroless/small attack surface for the purposes of security and compliance.

My original experiment request was here:
https://github.com/istio/istio/pull/12694#issuecomment-476124671

I think since the experiment seems to work (well enough) - the next step is to sort out how/if we  want to introduce a second set of distroless tuned dockerfiles into the repository.

File: install/kubernetes/helm/istio/charts/security/templates/tests/test-citadel-connection.yaml (hunk 7 above)
Comment (c0d1ngm0nk3y) [reply install/kubernetes/helm/istio/charts/security/templates/tests/test-citadel-connection.yaml]: > This doesn't look right? Is this supposed to be changed?

What exactly do you mean? This is only used for a curl for testing purposes. I don't see why this should use TAG/HUB.

Comment (sdake) [reply install/kubernetes/helm/istio/charts/security/templates/tests/test-citadel-connection.yaml]: @c0d1ngm0nk3y its a bit surprising to see.  A comment would help or perhaps what would make more sense is to break out a separate PR Rfor all of the testing charts and replace the images with the curl pod.

Cheers
-steve

**Comment by sdake** [comment]
> Would these images still contain libraries like `libnettle`, `libhogweed` etc?
> We are currently calling these from Lua filters in Istio because istio/proxy doesn't expose the embedded BoringSSL library.
> 
> If neither one nor the other is available, that would be a serious problem for us.

@robertpanzer re LUA filters and the need for other libraries, can you expand on which other libraries you would need or want?  I don't know how the LUA filters work and we want our operators to have a good user experience.  Taking away mandatory functionality is not that :)

Cheers
-steve

**Comment by sdake** [comment]
@utka and @istio/test-infra-hackers I think @costinm and I are in agreement - to be discussed at the April 4th Environments WG meeting on this feature for 1.2.  The next question that comes from this work is how do we test and release two sets of images.

I'd invite you to the environments wg meeting so we can all get on the same page as to why this work is needed and find a path forward.  The biggest impact I see is to the T&R workgroup.

Cheers
-steve

**Comment by robertpanzer** [comment]
> LUA filters and the need for other libraries, can you expand on which other libraries you would need or want? I don't know how the LUA filters work and we want our operators to have a good user experience. Taking away mandatory functionality is not that :)

Sure! :)
Currently our list contains:
- /usr/lib/x86_64-linux-gnu/libnettle.so.6
- /usr/lib/x86_64-linux-gnu/libhogweed.so.4
- /usr/lib/x86_64-linux-gnu/libgmp.so.10
- /lib/x86_64-linux-gnu/libc.so.6
- /lib/x86_64-linux-gnu/libuuid.so.1

**Comment by jsPlachetka** [comment]
Got merged into #13579.
